builder/builder.go

```diff
@@ -381,8 +401,22 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
 	return mts, nil
 }
 
+func (b *Builder) targetPhaseState(sts *states.SingleTarget) llb.State {
+	if b.builtMain {
+		return sts.RunPush.State
+	}
+	return sts.MainState
+}
+
+func (b *Builder) targetPhaseArtifacts(sts *states.SingleTarget) []states.SaveLocal {
+	if b.builtMain {
+		return sts.RunPush.SaveLocals
+	}
+	return sts.SaveLocals
+}
+
 func (b *Builder) stateToRef(ctx context.Context, gwClient gwclient.Client, state llb.State, platform *specs.Platform) (gwclient.Reference, error) {
-	if b.opt.NoCache {
+	if b.opt.NoCache || b.builtMain {
```

Comment:
This doesn't look right. I think it should be

```suggestion
	if b.opt.NoCache && !b.builtMain {
```

Meaning that for the push phase, we do not add the ignore cache flag here (instead we rely on the converter having added this flag to individual commands).

If we add the ignore cache flag here for the push phase, then everything the push phase depends on (including the main phase) will be executed with no cache - so the main phase ends up being executed twice.

Reply:
What you initially suggested is what I hade before; but I noticed that solving the states for the separate artifacts was showing up as cached. I used to have some `ls` sentinels in here to sho when things were being re-run, but I must have pulled them out before swapping to this.

I was trying to get `SAVE ARTIFACT` inside `--push` to _also_ not cache when running with `--no-cache`. Ill reqork this some to get both working